pyteal/ast/abi/util_test.py

```diff
@@ -771,3 +771,164 @@ def test_sdk_type_specs_from_signature(sig_str, sig_args, sig_rets):
+        False,
+    ),
+    (
+        abi.type_spec_from_annotation(NamedTDecl),
+        abi.type_spec_from_annotation(
+            abi.Tuple3[
+                abi.Uint64,
+                abi.Tuple3[
+                    abi.Transaction, abi.Address, abi.StaticArray[abi.Byte, Literal[16]]
+                ],
+                abi.Transaction,
+            ]
+        ),
+        True,
+    ),
+    (
+        abi.type_spec_from_annotation(
+            abi.Tuple3[
+                abi.Uint64,
+                abi.Tuple3[
+                    abi.PaymentTransaction,
+                    abi.Address,
+                    abi.StaticArray[abi.Byte, Literal[16]],
+                ],
+                abi.Transaction,
+            ]
+        ),
+        abi.type_spec_from_annotation(NamedTDecl),
+        True,
+    ),
+    (
+        abi.type_spec_from_annotation(abi.StaticBytes[Literal[7]]),
+        abi.type_spec_from_annotation(abi.StaticArray[abi.Byte, Literal[11]]),
+        False,
+    ),
+    (
+        abi.type_spec_from_annotation(NamedTDecl),
+        abi.type_spec_from_annotation(NamedTDecl),
+        True,
+    ),
+    (
+        abi.type_spec_from_annotation(abi.String),
+        abi.type_spec_from_annotation(abi.DynamicBytes),
+        True,
+    ),
+    (
+        abi.type_spec_from_annotation(abi.DynamicArray[abi.Byte]),
+        abi.type_spec_from_annotation(abi.String),
+        True,
+    ),
+    (
+        abi.type_spec_from_annotation(abi.DynamicArray[abi.Uint32]),
+        abi.type_spec_from_annotation(abi.String),
+        False,
+    ),
+    (
+        abi.type_spec_from_annotation(abi.Address),
+        abi.type_spec_from_annotation(abi.StaticArray[abi.Byte, Literal[32]]),
+        True,
+    ),
+    (
+        abi.type_spec_from_annotation(abi.StaticBytes[Literal[32]]),
+        abi.type_spec_from_annotation(abi.Address),
+        True,
+    ),
+    (
+        abi.type_spec_from_annotation(abi.StaticBytes[Literal[33]]),
+        abi.type_spec_from_annotation(abi.Address),
+        False,
+    ),
+    (
+        abi.type_spec_from_annotation(NamedTComp0),
+        abi.type_spec_from_annotation(NamedTComp1),
+        True,
+    ),
+    (
+        abi.type_spec_from_annotation(NamedTComp1),
+        abi.type_spec_from_annotation(NamedTComp0),
+        True,
+    ),
+    (
+        abi.type_spec_from_annotation(NamedTDecl),
+        abi.type_spec_from_annotation(NamedTComp0),
+        False,
+    ),
+    (
+        abi.type_spec_from_annotation(NamedTComp2),
+        abi.type_spec_from_annotation(NamedTComp0),
+        False,
+    ),
+    (
+        abi.type_spec_from_annotation(abi.String),
+        abi.type_spec_from_annotation(abi.DynamicArray[abi.Uint64]),
+        False,
+    ),
+    (
+        abi.type_spec_from_annotation(abi.DynamicArray[abi.Uint64]),
+        abi.type_spec_from_annotation(abi.String),
+        False,
+    ),
+]
+
+
+@pytest.mark.parametrize("a, b, expected", TYPE_SPEC_ASSIGNABLE_CASES)
```

Comment:
Leaving a note to document a verbal with @ahangsu and @jasonpaulos:  We discussed + agreed it'd make sense to enforce a more exhaustive exercising of `TypeSpec` assignability.  

* Rationale:  Without having to rely on a maintainer's memory alone to extend test cases, we want to defend against regressions as new `TypeSepc`s are introduced / modified.
* Caveats:  
  * We did _not_ agree upon an implementation approach.  It's understood that @ahangsu will make a pass and we can revisit.
  * There's some open questions around ease to exhaustively check _constructive_ types (e.g. `TupleTypeSpec`, `ArrayTypeSpec`).  It's understood that the solution may _not_ fully automate assignability checks.
* Related: In case helpful for brainstorming, we recently implemented a safe guard to warn us that new test cases are needed in https://github.com/algorand/pyteal/pull/519/files#diff-a1205be6b51287e45f2be7311682488d51bddf8900c13deb97df282aa090349bR396.  It's _not_ a direct fit to the problem at-hand, but it may provide inspiration.